components/suggest/src/store.rs

```diff
@@ -366,31 +366,37 @@ where
             return Ok(());
         };
 
-        let attachment: SuggestAttachment<T> =
-            serde_json::from_slice(&self.settings_client.get_attachment(&attachment.location)?)?;
-
-        writer.write(|dao| {
-            // Drop any suggestions that we previously ingested from
-            // this record's attachment. Suggestions don't have a
-            // stable identifier, and determining which suggestions in
-            // the attachment actually changed is more complicated than
-            // dropping and re-ingesting all of them.
-            dao.drop_suggestions(&record_id)?;
-
-            // Ingest (or re-ingest) all suggestions in the
-            // attachment.
-            ingestion_handler(dao, &record_id, attachment.suggestions())?;
-
-            // Remove this record's ID from the list of unparsable
-            // records, since we understand it now.
-            dao.drop_unparsable_record_id(&record_id)?;
-
-            // Advance the last fetch time, so that we can resume
-            // fetching after this record if we're interrupted.
-            dao.put_last_ingest_if_newer(record.last_modified)?;
-
-            Ok(())
-        })
+        let attachment_data = self.settings_client.get_attachment(&attachment.location)?;
+        match serde_json::from_slice::<SuggestAttachment<T>>(&attachment_data) {
+            Ok(attachment) => writer.write(|dao| {
+                // Drop any suggestions that we previously ingested from
+                // this record's attachment. Suggestions don't have a
+                // stable identifier, and determining which suggestions in
+                // the attachment actually changed is more complicated than
+                // dropping and re-ingesting all of them.
+                dao.drop_suggestions(&record_id)?;
+
+                // Ingest (or re-ingest) all suggestions in the
+                // attachment.
+                ingestion_handler(dao, &record_id, attachment.suggestions())?;
+
+                // Remove this record's ID from the list of unparsable
+                // records, since we understand it now.
+                dao.drop_unparsable_record_id(&record_id)?;
+
+                // Advance the last fetch time, so that we can resume
+                // fetching after this record if we're interrupted.
+                dao.put_last_ingest_if_newer(record.last_modified)?;
+
+                Ok(())
+            }),
+            Err(_) => writer.write(|dao| {
+                // The attachment was not able to be parsed, record this and continue on
+                dao.put_unparsable_record_id(&record_id)?;
+                dao.put_last_ingest_if_newer(record.last_modified)?;
```

Comment:
Would it make sense to put this code in a `dao` method?  I slightly worry that somewhere down the line there will be code that only does the first operation, but not the second one.  Alternatively, maybe `SuggestDao` only have low-level CRUD methods and a higher-level component could handle compound operations like this one.

Reply:
I think it would make sense, and `SuggestDao` sounds like a good place to put it! If it starts getting unwieldy having the lower-level methods and the higher-level ones in `SuggestDao`, maybe we could think about separating them down the road, but I'm not sure we've reached that point just yet.

Our other access pattern is to do `drop_unparsable_record_id + put_last_ingest_if_newer` for records that we did parse and store successfully—worth having a second `SuggestDao` helper for that, too?